log/field.go

```diff
@@ -0,0 +1,201 @@
+package log
+
+import "math"
+
+type fieldType int
+
+const (
+	stringType fieldType = iota
+	boolType
+	intType
+	int32Type
+	uint32Type
+	int64Type
+	uint64Type
+	float32Type
+	float64Type
+	errorType
+	objectType
+	lazyLoggerType
```

Comment:
Personally, I see a strong enough use case for logging structs (as `interface{}` values, via something like the `log.Object(...)` in this PR), but pretty much as just a big (possibly truncated) blob for debugging purposes. Nested key-value fields at this logging layer are not indefensible, but IMO cross the complexity threshold and aren't worth the corner cases they expose.

And there's the perennial "easier to add than to remove" argument, too...